### PR TITLE
Handle empty block

### PR DESCRIPTION
**Pull request by lucas7788**

### CODE REVIEW

File: p2pserver/block_sync.go [hunk 1]
```diff
@@ -277,34 +274,34 @@ func (this *BlockCache) addBlockCache(nodeID uint64, block *types.Block,
...
 	}
 	return blockInfo.nodeID, blockInfo.block, blockInfo.merkleRoot
 }
 
-func (this *BlockCache) delBlockCache(blockHeight uint32) {
-	this.lock.Lock()
-	defer this.lock.Unlock()
-	this.delBlock(blockHeight)
+func (this *BlockCache) delBlock(blockHeight uint32) {
+	this.delBlockLocked(blockHeight)
```
Comment (laizy) [inline p2pserver/block_sync.go]: those two functions are equal now, remove one

File: p2pserver/block_sync.go [hunk 2]
```diff
@@ -665,16 +752,7 @@ func (this *BlockSyncMgr) saveBlock() {
...
-	defer this.lock.RUnlock()
-	_, ok := this.blocksCache[blockHeight]
-	return ok
-}
-
-func (this *BlockSyncMgr) getBlockCacheSize() int {
-	this.lock.RLock()
-	defer this.lock.RUnlock()
-	return len(this.blocksCache)
+	return this.blocksCache.isInBlockCache(blockHeight)
```
Comment (laizy) [inline p2pserver/block_sync.go:754]: need lock here

File: p2pserver/block_sync.go [hunk 3]
```diff
@@ -240,14 +240,87 @@ func NewBlockSyncMgr(server *P2PServer) *BlockSyncMgr {
...
 		server:        server,
 		ledger:        server.ledger,
 		exitCh:        make(chan interface{}, 1),
 		nodeWeights:   make(map[uint64]*NodeWeight, 0),
 	}
 }
 
+type BlockCache struct {
+	emptyBlockAmount int
+	blocksCache      map[uint32]*BlockInfo //Map BlockHash => BlockInfo, using for cache the blocks receive from net, and waiting for commit to ledger
```
Comment (laizy) [inline p2pserver/block_sync.go]:
```suggestion
	blocksCache      map[uint32]*BlockInfo //Map BlockHeight => BlockInfo, using for cache the blocks receive from net, and waiting for commit to ledger
```

File: p2pserver/block_sync.go [hunk 4]
```diff
@@ -240,14 +240,87 @@ func NewBlockSyncMgr(server *P2PServer) *BlockSyncMgr {
...
+func (this *BlockCache) addBlock(nodeID uint64, block *types.Block,
+	merkleRoot common.Uint256) bool {
+	blockInfo := &BlockInfo{
+		nodeID:     nodeID,
+		block:      block,
+		merkleRoot: merkleRoot,
+	}
+	this.blocksCache[block.Header.Height] = blockInfo
+	if block.Header.TransactionsRoot == common.UINT256_EMPTY {
+		this.emptyBlockAmount += 1
```
Comment (laizy) [inline p2pserver/block_sync.go:273]: check block existence

**Review by laizy — DISMISSED** [review]

File: p2pserver/block_sync.go [hunk 5]
```diff
@@ -240,14 +240,88 @@ func NewBlockSyncMgr(server *P2PServer) *BlockSyncMgr {
...
+	}
+	delete(this.blocksCache, blockHeight)
+}
+
+func (this *BlockCache) isInBlockCache(blockHeight uint32) bool {
+	_, ok := this.blocksCache[blockHeight]
+	return ok
+}
+
+func (this *BlockCache) getNonEmptyBlockSize() int {
```
Comment (Honglei-Cong) [inline p2pserver/block_sync.go]: rename as `getNonEmptyBlockCount()` ?

File: p2pserver/block_sync.go [hunk 6]
```diff
@@ -240,14 +240,88 @@ func NewBlockSyncMgr(server *P2PServer) *BlockSyncMgr {
...
+	return len(this.blocksCache) - this.emptyBlockAmount
+}
+func (this *BlockSyncMgr) getNonEmptyBlockSize() int {
```
Comment (Honglei-Cong) [inline p2pserver/block_sync.go]: ditto

**Review by Honglei-Cong — APPROVED** [review]

**Pull request merged**